dual_net.py

```diff
@@ -301,61 +312,39 @@ def logging_ops():
         name="weight_summaries")
 
 
+def compute_gradient_ratio(weight_tensors, before_weights, after_weights):
+    """Compute the ratio of gradient norm to weight norm."""
+    deltas = [after - before for after, before in zip(after_weights, before_weights)]
+    delta_norms = [np.linalg.norm(d.ravel()) for d in deltas]
+    weight_norms = [np.linalg.norm(w.ravel()) for w in before_weights]
+    ratios = [d / w for d, w in zip(delta_norms, weight_norms)]
+    all_summaries = [tf.Summary.Value(tag=tensor.name, simple_value=ratio)
+        for tensor, ratio in zip(weight_tensors, ratios)]
+    return tf.Summary(value=all_summaries)
+
+
 class StatisticsCollector(object):
     """Collect statistics on the runs and create graphs.
 
     Accuracy and cost cannot be calculated with the full test dataset
     in one pass, so they must be computed in batches. Unfortunately,
     the built-in TF summary nodes cannot be told to aggregate multiple
     executions. Therefore, we aggregate the accuracy/cost ourselves at
-    the python level, and then shove it through the accuracy/cost summary
-    nodes to generate the appropriate summary protobufs for writing.
+    the python level, and then generate the summary protobufs for writing.
     """
-    graph = tf.Graph()
-    with tf.device("/cpu:0"), graph.as_default():
-        policy_error = tf.placeholder(tf.float32, [])
-        value_error = tf.placeholder(tf.float32, [])
-        reg_error = tf.placeholder(tf.float32, [])
-        cost = tf.placeholder(tf.float32, [])
-        policy_summary = tf.summary.scalar("Policy error", policy_error)
-        value_summary = tf.summary.scalar("Value error", value_error)
-        reg_summary = tf.summary.scalar("Regularization error", reg_error)
-        cost_summary = tf.summary.scalar("Combined cost", cost)
-        accuracy_summaries = tf.summary.merge(
-            [policy_summary, value_summary, reg_summary, cost_summary],
-            name="accuracy_summaries")
-    config = tf.ConfigProto()
-    config.gpu_options.allow_growth = True
-    session = tf.Session(graph=tf.Graph(), config=config)
-
     def __init__(self):
-        self.policy_costs = []
-        self.value_costs = []
-        self.regularization_costs = []
-        self.combined_costs = []
-
-    def report(self, policy_cost, value_cost, regularization_cost, combined_cost):
-        # TODO refactor this to take a dict, and do something like
-        # self.accums = defaultdict(list) and do self.accums[thing].append(value)
-        # so that it can handle an arbitrary number of values.
-        self.policy_costs.append(policy_cost)
-        self.value_costs.append(value_cost)
-        self.regularization_costs.append(regularization_cost)
-        self.combined_costs.append(combined_cost)
+        self.accums = collections.defaultdict(list)
+
+    def report(self, values):
+        """Take a dict of scalar names to scalars, and aggregate by name."""
+        for key, val in values.items():
```

Comment:
can this be a comprehension?  [self.accums[key].append(val) for key,val in values.items()] ?

Reply:
a comprehension that has side effects like appending feels weird to me.

Reply:
no problem.  LGTM then.